ethpm/tools/builder.py

```diff
@@ -328,9 +321,86 @@ def normalize_contract_type(
     """
     Serialize contract_data found in compiler output to the defined fields.
     """
-    yield "abi", json.loads(contract_type_data["abi"])
-    yield "deployment_bytecode", {"bytecode": add_0x_prefix(contract_type_data["bin"])}
-    yield "natspec", json.loads(contract_type_data["devdoc"])
+    yield "abi", contract_type_data["abi"]
+    yield "deployment_bytecode", normalize_bytecode_object(
+        contract_type_data["evm"]["bytecode"]
+    )
+    # todo support natspec, runtime_bytecode
+    yield "natspec", {}
+
+
+@to_dict
+def normalize_bytecode_object(
+    obj: Dict[str, Any]
+) -> Generator[Dict[str, Any], None, None]:
+    if obj["linkReferences"]:
+        yield "link_references", process_link_references(
+            obj["linkReferences"], obj["object"]
+        )
+        yield "bytecode", process_bytecode(obj["linkReferences"], obj["object"])
+    else:
+        yield "bytecode", add_0x_prefix(obj["object"])
+
+
+def process_bytecode(link_refs: Dict[str, Any], bytecode: bytes) -> bytes:
+    """
+    Replace link_refs in bytecode with 0's.
+    """
+    all_offsets = [y for x in link_refs.values() for y in x.values()]
+    link_fns = (
+        replace_link_ref_in_bytecode(ref["start"] * 2, ref["length"] * 2)
+        for ref in list(itertools.chain.from_iterable(all_offsets))
+    )
+    processed_bytecode = cytoolz.pipe(bytecode, *link_fns)
+    return add_0x_prefix(processed_bytecode)
+
+
+@cytoolz.curry
+def replace_link_ref_in_bytecode(offset: int, length: int, bytecode: bytes) -> bytes:
+    validate_link_ref(offset, length, bytecode)
+    new_bytes = (
+        bytecode[:offset] + "0" * length + bytecode[offset + length :]  # noqa: E203
+    )
+    return new_bytes
+
+
+# todo pull all bytecode linking/validating across py-ethpm into shared utils
+@to_list
+def process_link_references(
+    link_refs: Dict[str, Any], bytecode: bytes
+) -> Generator[List[Dict[str, Any]], None, None]:
+    for link_ref in link_refs.values():
+        yield normalize_link_ref(link_ref, bytecode)
+
+
+@to_dict
+def normalize_link_ref(
+    link_ref: Dict[str, Any], bytecode: bytes
+) -> Generator[Dict[str, Any], None, None]:
+    for name, real_data in link_ref.items():
+        yield "name", name
+        yield "length", 20
+        yield "offsets", normalize_offsets(link_ref, bytecode)
```

Comment:
This whole function is confusing.  The `to_dict` implies that it is making a *single* dictionary.  The type hinting implies it is supposed to make many dictionaries.  I think this body inside the `for` loop should be yielding a single dictionary rather than the key/value format.  If you run this function I think you'll find it will only ever produce a single link reference rather than the iterable of them you are probably expecting.